src/Data/Expr.purs

```diff
@@ -4,259 +4,157 @@ import Prelude
-        e :: ExprF Boolean Int <- decodeJson json
-        pure $ mkExpr e
-      "Int" -> do
-        e :: ExprF Int Int <- decodeJson json
-        pure $ mkExpr e
-      "String" -> do
-        e :: ExprF String Int <- decodeJson json
-        pure $ mkExpr e
-      "Void" -> do
-        e :: ExprF Void Int <- decodeJson json
-        pure $ mkExpr e
-      other -> Left $ other <> " not implemented"
-else instance decodeExprString :: DecodeJson (Expr String) where
-  decodeJson json = do
-    x <- decodeJson json
-    x .: "in" >>= case _ of
-      "Boolean" -> do
-        e :: ExprF Boolean String <- decodeJson json
-        pure $ mkExpr e
-      "Int" -> do
-        e :: ExprF Int String <- decodeJson json
-        pure $ mkExpr e
-      "String" -> do
-        e :: ExprF String String <- decodeJson json
-        pure $ mkExpr e
-      "Void" -> do
-        e :: ExprF Void String <- decodeJson json
-        pure $ mkExpr e
-      other -> Left $ other <> " not implemented"
-
-instance decodeVal :: DecodeJson o => DecodeJson (ExprF Void o) where
-  decodeJson json = do
-    x <- decodeJson json
-    x .: "param" >>= pure <<< Val identity
-else instance decodeExprFBooleanBoolean :: DecodeJson (ExprF Boolean Boolean) where
-  decodeJson json = do
-    x <- decodeJson json
-    params <- x .: "params"
-    x .: "op" >>= case _ of
-      "If" -> decodeIfF params
-      "Equal" -> decodeEqualF params
-      op -> Left $ op <> " invald op"
-else instance decodeExprFIBoolean :: DecodeJson (Expr i) => DecodeJson (ExprF i Boolean) where
-  decodeJson json = do
-    x <- decodeJson json
-    params <- x .: "params"
-    x .: "op" >>= case _ of
-      "Equal" -> decodeEqualF params
-      op -> Left $ op <> " invalid op"
-else instance decodeExprFBooleanO :: DecodeJson (Expr o) => DecodeJson (ExprF Boolean o) where
-  decodeJson json = do
-    x <- decodeJson json
-    params <- x .: "params"
-    x .: "op" >>= case _ of
-      "If" -> decodeIfF params
-      op -> Left $ op <> " invalid op"
-else instance decodeExprFOther :: DecodeJson (ExprF i o) where
-  decodeJson _ = Left "Unimplemented"
-
-decodeIfF
-  :: ∀ o
-   . DecodeJson (Expr o)
-  => Array Json
-  -> Either String (ExprF Boolean o)
-decodeIfF [ x, y, z ] = do
-  x' :: Expr Boolean <- decodeJson x
-  y' :: Expr o <- decodeJson y
-  z' :: Expr o <- decodeJson z
-  pure $ If identity x' y' z'
-decodeIfF xs = Left "Expceted 3 params"
-
-decodeEqualF
-  :: ∀ i
-   . DecodeJson (Expr i)
-  => Array Json
-  -> Either String (ExprF i Boolean)
-decodeEqualF [ x, y ] = do
-  x' :: Expr i <- decodeJson x
-  y' :: Expr i <- decodeJson y
-  pure $ Equal identity x' y'
-decodeEqualF xs = Left "Expected 2 params"
+data EvalError
+  = MissingKey Key
+  | IfCondition ExprType
+  | EqualMismatch { left :: ExprType, right :: ExprType }
+
+evalExpr :: (Key -> Maybe ExprType) -> Expr -> Either EvalError ExprType
+evalExpr get = case _ of
+  Val x -> pure x
+  If x' y z -> case evalExpr get x' of
```

Comment:
I think the explicit casing on the `Either _ _` can be replaced with binds. Maybe something like:

```PureScript
If x' y z -> do
  x <- evalExpr get x'
  case x of
    Boolean false -> evalExpr get z
    Boolean true -> evalExpr get y
    _ -> Left (IfCondition x)
```
or
```PureScript
If x' y z -> evalExpr get x' >>= case _ of
  Boolean false -> evalExpr get z
  Boolean true -> evalExpr get y
  x -> Left (IfCondition x)
```

Not a huge deal on this one, but it might clean up the `Equal` case a bit.